src/job.rs

```diff
@@ -127,31 +144,62 @@ impl Job {
                 key: hasher.finish(),
                 phantom: PhantomData,
             },
-            env: unwrapped.env,
-            command: unwrapped.command,
+            env: &unwrapped.env,
+            command: &unwrapped.command,
             input_files,
+            input_jobs,
             outputs,
         })
     }
+
+    pub fn final_key(
+        &self,
+        path_to_hash: &HashMap<PathBuf, String>,
+        job_to_content_hash: &HashMap<Key<Base>, store::Item>,
+    ) -> Result<Key<Final>> {
+        let mut hasher = Xxh3::new();
+
+        self.base_key.hash(&mut hasher);
+
+        for path in &self.input_files {
+            match path_to_hash.get(path) {
+                Some(hash) => {
+                    path.hash(&mut hasher);
+                    hash.hash(&mut hasher);
+                },
+                None => anyhow::bail!("`{}` was specified as a file dependency, but I didn't have a hash for it! This is a bug in rbt's coordinator, please file it!", path.display()),
+            }
+        }
+
+        for key in self.input_jobs.keys().sorted() {
+            let dep = job_to_content_hash.get(key).context("could not look up output hash for dependency. This is a bug in rbt's coordinator. Please file it!")?.hash();
```

Comment:
What hasher does this `hash()` use?

Reply:
it's a getter for content-addressable store key, whose hashing has already been done. (It uses blake3 there.)